driver/apdu/mbim.c

```diff
@@ -0,0 +1,353 @@
+    guint32 response_size = 0;
+    const guint8 *response_data = NULL;
+
+    if (!mbim_message_ms_uicc_low_level_access_open_channel_response_parse(
+        response, &status, &channel, &response_size, &response_data, &error
+    )) {
+        fprintf(stderr, "error: unable to parse channel response: %s\n", error->message);
+        return -1;
+    }
+
+    mbim_priv->last_channel_id = channel;
+    return channel;
+}
+
+static void mbim_apdu_interface_logic_channel_close(struct euicc_ctx *ctx, uint8_t channel)
+{
+    struct mbim_data *mbim_priv = ctx->apdu.interface->userdata;
+    g_autoptr(GError) error = NULL;
+
+    MbimMessage *request = mbim_message_ms_uicc_low_level_access_close_channel_set_new(
+        channel, 1, &error
+    );
+    if (!request) {
+        fprintf(stderr, "error: creating channel message failed: %s\n", error->message);
+        return;
+    }
+
+    g_autoptr(MbimMessage) response = mbim_device_command_sync(
+        mbim_priv->device, mbim_priv->context, request, &error
+    );
+    if (!response) {
+        fprintf(stderr, "error: no channel response received: %s\n", error->message);
+        return;
+    }
+
+    guint32 status = 0;
+
+    if (!mbim_message_ms_uicc_low_level_access_close_channel_response_parse(
+        response, &status, &error
+    )) {
+        fprintf(stderr, "error: unable to parse channel response: %s\n", error->message);
+        return;
+    }
+
+    if (channel == mbim_priv->last_channel_id)
+        mbim_priv->last_channel_id = -1;
+
+    g_debug("Closed logical channel with id %d", channel);
+}
+
+static void mbim_apdu_interface_disconnect(struct euicc_ctx *ctx)
+{
+    struct mbim_data *mbim_priv = ctx->apdu.interface->userdata;
+    g_autoptr(GError) error = NULL;
+
+    if (mbim_priv->last_channel_id > 0)
+    {
+        fprintf(stderr, "Cleaning up leaked APDU channel %d\n", mbim_priv->last_channel_id);
+        mbim_apdu_interface_logic_channel_close(NULL, mbim_priv->last_channel_id);
+        mbim_priv->last_channel_id = -1;
+    }
+
+    mbim_device_close_sync(mbim_priv->device, mbim_priv->context, &error);
+
+    g_main_context_unref(mbim_priv->context);
+    mbim_priv->context = NULL;
+}
+
+static int libapduinterface_init(struct euicc_apdu_interface *ifstruct)
+{
+    struct mbim_data *mbim_priv;
+
+    mbim_priv = malloc(sizeof(struct mbim_data));
+    if (!mbim_priv) {
+        fprintf(stderr, "Failed allocating memory\n");
+        return -1;
+    }
+
+    memset(ifstruct, 0, sizeof(struct euicc_apdu_interface));
+    ifstruct->connect = apdu_interface_connect;
+    ifstruct->disconnect = mbim_apdu_interface_disconnect;
+    ifstruct->logic_channel_open = mbim_apdu_interface_logic_channel_open;
+    ifstruct->logic_channel_close = mbim_apdu_interface_logic_channel_close;
+    ifstruct->transmit = mbim_apdu_interface_transmit;
+    ifstruct->userdata = mbim_priv;
+
+    const char *uim_slot = getenv("UIM_SLOT");
+    mbim_priv->uim_slot = uim_slot ? atoi(uim_slot) : 1;
```

Comment:
> 0 is a valid uim selection though

But `UIM_SLOT=0` is different with `UIM_SLOT=somegarbagestring`, treat the later one as same as the the former doesn't make sense. If someone, or some program invoke lpac with garbage input, but got the behavior of `UIM_SLOT=0`. I don't think it's expected and safe.

Reply:
As I said, I don't disagree. Qmi needs to see this fixed as well.

Reply:
> As I said, I don't disagree. Qmi needs to see this fixed as well.

Sorry, I misread it as "I don't agree". ):

Reply:
So what kind of behavior would you want here? Both sscanf and atoi are going to produce 0 as a result when we're starting with a non-digit, with the advantage for sscanf being that we can immediately see that it did that. However, what behavior would you expect when the environment variable is set to "1 ", or "1st slot"? Would you consider that valid input or not? Either case is going to produce 1 as a result from both functions. And "1st slot" is possibly going to produce a result that the user isn't expecting either.

Reply:
This is what I want:
1. If UIM_SLOT is a string form of a valid number, convert it to a number, and then
   1. If the number is a valid slot index, use it.
   2. Otherwise, raise an error and exit.
2. Otherwise, raise an error and exit.

Reply:
OK. I have yet to run into modems that have more than two sim slots available. I'd go as far as saying that it is unlikely that we're going into the double digits any time soon. The slot selection string can then be required to be a single character digit.

I don't know up-front if the modem has one or two slots (or more), so we'll have to run into that when trying to switch, and error and exit there.

Reply:
Looks like qmi slot selection is 1-based and it kind of accidentally handles faulty input. Since I'm hijacking the same UIM_SLOT variable, I'll stick to the 1-based approach in this interface as well for consistency.